Add CodeMirror to the HTML block

noisysocks opened the pull request:
### 👀 What this is

Adds CodeMirror to the HTML block. This means that a user can type their custom HTML using a fully featured text editor with linting and syntax highlighting:

![codemirror](https://user-images.githubusercontent.com/612155/34701754-10693cbe-f53e-11e7-9ca6-5c98bf4d6200.gif)

### 🤔 How this works

[CodeMirror was added to core in WordPress 4.9](https://make.wordpress.org/core/2017/10/22/code-editing-improvements-in-wordpress-4-9/) which means we can enhance a `<textarea>` simply by calling `wp.codeEditor.initialize()`.

To make CodeMirror easier to use within React, I introduced a `<CodeEditor>` component which acts as a drop-in replacement for a `<textarea>`.

I've also incorporated some of @westonruter's work in https://github.com/WordPress/gutenberg/pull/1625 by using `SandBox` within our HTML block. This makes previewing a HTML block more robust. Specifically, you can now insert HTML that blocks rendering, e.g.:

```html
<script>document.write( 'Hey!' );</script>
```

To avoiding adding a whopping 1.9 MB to our assets, I've implemented a very simple mechanism to lazily load the assets that are normally queued by calling `wp_enqueue_code_editor()`. 

### ✅ How to test

1. Create a HTML block
2. Preview it
3. Save it, view the page
4. Check that you can preview HTML that blocks rendering, e.g. `<script>document.write( 'Hey!' );</script>`
4. Log in as an author and test that you get lint warnings when using e.g. the `<script>` tag

noisysocks committed:
Use CodeMirror in Custom HTML block

Introduces a new <CodeEditor> component which wraps a CodeMirror
enhanced <textarea>. This component is then used to provide syntax
highliting in the Custom HTML block.

noisysocks committed:
Prevent CodeEditor from propagating events

This prevents the cursor from moving around to different blocks while
editing code.

noisysocks committed:
Allow CodeEditor to be focused via the `focus` prop

gziolo commented:
Cross-posting related issue: #2768. I didn't take any measurements on the size of CodeMirror, but I would expect it isn't a small one :) I think it's fine to include it temporarily on the initial editor load. However we should at least double check what impact it has on the size of JS and CSS loaded, and mention it in the #2768 so we know if we have to act before it gets merged into the core.

noisysocks commented:
Here are some quick measurements from running `npm run build` and checking DevTools:

| Branch | All | JS | CSS |
| --- | --- | --- | --- |
| `master` | 3.4 MB | 2.9 MB | 261 KB |
| This PR | 5.3 MB | 4.8 MB | 277 KB |

Yikes! A 1.9 MB (55%) increase! 😱 I'll definitely look into dynamically loading the assets.

noisysocks commented:
cc'ing some people who might want to track this work:

- @mtias — suggested it
- @jasmussen — mocked it up in #1386 
- @westonruter — worked on adding CodeMirror to WP Core
- @dmsnell — worked on adding a CodeMirror block to Gutenberg in #1625

noisysocks committed:
Improve HTML block styling

Make the CodeEditor in the HTML block look vaguely like what Joen mocked
up in #1386.

noisysocks committed:
Allow HTML block to be selected when it's in preview mode

Prevent the iframe from capturing pointer events and preventing the user
from being able to select the block.

noisysocks committed:
Always focus the editor after React has finished painting

`componentDidUpdate` and `componentDidMount` fire after DOM
reconcilation but before the frame is painted. CodeMirror's `focus()`
method seems to only work *after* paint. Scheduling `focus()` to happen
after the next frame is painted seems to be the best that we can do.

gziolo commented:
> A 1.9 MB (55%) increase! 😱 

@noisysocks thanks for checking 🙇 

Another thing we can add as a follow-up PR is HTML linting which is built-in in the core. By the way, the code for linting is lazy-loaded when editing plugin files. Which is a very good sign since we discovered that CodeMirror adds 1.9 MB to the uncompressed build size.

gziolo commented on components/code-editor/test/index.js:
```diff
@@ -0,0 +1,23 @@
+		} ) );
+
+		const wrapper = shallow( <CodeEditor value={ '<b>wowee</b>' } /> );
+		expect( wrapper.find( 'textarea' ) ).toHaveLength( 1 );
```
We could use snapshot testing here, too. See example: https://github.com/WordPress/gutenberg/blob/300ca0a8acc73ff4013c0495a8c636527fcf588f/components/notice/test/index.js#L15 and docs: https://github.com/WordPress/gutenberg/blob/master/docs/testing-overview.md#snapshot-testing.

noisysocks replied:
Good idea. I've changed this to a snapshot test in 17dbeb231a901007b08e5fd0dd75840bdbbc1489.

mtias commented:
> A 1.9 MB (55%) increase! 😱

Ouch.

noisysocks committed:
Change CodeEditor test to use a snapshot test

We only care that the component renders without an error and doesn't
change unintentionally, which makes this a perfect use case for snapshot
testing.

noisysocks committed:
Lazily load CodeMirror assets

Wraps CodeEditor with a component that lazily loads the scripts and
styles that CodeMirror needs. This lets us avoid using
wp_enqueue_code_editor() which adds considerable bulk (~ 1.9 MB) to the
page load.

noisysocks commented:
Any feedback, questions, or concerns with this PR?

I'd like to get the basic functionality merged and then add lazy loading in a seperate PR. I've started this work over in https://github.com/WordPress/gutenberg/pull/4500.

jasmussen commented:
Really nice, work, fun to see this, it's a very very cool block.

There are a couple of things we should address in this PR. First off, can we tweak the padding a little bit so it looks like this?

<img width="745" alt="screen shot 2018-01-16 at 13 36 08" src="https://user-images.githubusercontent.com/1204802/34989286-4be8f1d8-fac2-11e7-8e68-4a2267757adb.png">

I think this should do it:

```
.CodeMirror-lines {
padding: 8px 8px 8px 0;
}
```

There's a z-index issue with adjecent block toolbars:

<img width="343" alt="screen shot 2018-01-16 at 13 33 12" src="https://user-images.githubusercontent.com/1204802/34989327-6af59630-fac2-11e7-9fda-38c6703abbfa.png">

One important issue we have to fix is that you need to be able to escape this block using just the keyboard:

- Pressing Escape while you're editing inside the block should deselect the entire block, just like it does in a Paragraph block
- Pressing the ↓ at the end of a codemirror block should move the caret into the next adjecent block, same with ↑ at the beginning of a block
- _Ideally_ you'd also be able to press Enter twice at the end of a codemirror block to create a new Paragraph block below, just like how you can press Enter twice inside a list to exit it

Question for a future PR, doesn't have to be this one — can we customize the colors of the syntax highlighting? If yes, I'd love for us to use more WordPressy colors. Alternatively, I'd love it if we could at least use something like `$blue-medium-200` for the line-highlighter.

noisysocks commented:
> Question for a future PR, doesn't have to be this one — can we customize the colors of the syntax highlighting?

Yep! It's just a matter of styling the `.CodeMirror` classes 🙂

noisysocks committed:
Increase code editor padding in HTML block

noisysocks committed:
Increase z-index of block toolbar elements

Bumps up the z-index of every element that is at the same depth as the
block toolbar. By making these indicies greater than 4, we ensure our
contextual elements appear over the CodeMirror editor. This stops the
CodeMirror gutter from appearing over the block toolbar.

noisysocks committed:
Improve HTML block cursor behaviour

- Permit pressing ESCAPE to unfocus the currently selected block.
- Allow pressing UP/DOWN to move focus to the previous/next block when
  the cursor is at the start/end of the editor.

noisysocks commented:
Thanks @jasmussen! I've implemented your suggested changes.

> Pressing the ↓ at the end of a codemirror block should move the caret into the next adjecent block, same with ↑ at the beginning of a block

I made it so that pressing UP/DOWN first moves the cursor to the start/end of the line. A subsequent press will then move focus to the previous/next block. It feels really natural in my testing.

> Ideally you'd also be able to press Enter twice at the end of a codemirror block to create a new Paragraph block below, just like how you can press Enter twice inside a list to exit it

I didn't implement this since pressing ENTER twice is quite common while writing code and e.g. adding whitespace between lines.

dmsnell commented:
sorry I'm coming in somewhat late. what did we decide about async-loading CodeMirror? are we wanting to still do that?

westonruter commented:
> sorry I'm coming in somewhat late. what did we decide about async-loading CodeMirror? are we wanting to still do that?

Yes, see https://github.com/WordPress/gutenberg/pull/4500

noisysocks commented:
Any more thoughts on this? 😊

dmsnell commented:
I'd really like to see us wait until this is loading on-demand before merging it otherwise it's going to instantly bloat the editor in no small way - that is, I'd love to see something like #4500 merge first _or_ have some changes here to lazy-load.

noisysocks commented:
I'll hold off on hitting the merge button until #4500 is ready. Nonetheless, I'd like to get this PR into an approved and ready-to-go state 🙂

jasmussen commented:
Visually this is in a good place 👍:

<img width="785" alt="screen shot 2018-01-23 at 09 05 26" src="https://user-images.githubusercontent.com/1204802/35264320-96612596-001c-11e8-9b6e-91a517aa675b.png">

I also appreciate that the arrow at the beginning and end of the block enters adjecent blocks or creates a new paragraph so at least you can escape the block. 👍 

However I'm seeing some weird behavior when testing. Not sure if it's intentional, but basic dummy text paragraphs are joined together on reload:

![codeblock](https://user-images.githubusercontent.com/1204802/35264425-fb3aa0d2-001c-11e8-9345-c7f17aa24b5b.gif)

noisysocks commented:
> However I'm seeing some weird behavior when testing. Not sure if it's intentional, but basic dummy text paragraphs are joined together on reload:

Good catch! Looks like this is because Gutenberg pretty prints the post HTML before saving. Since this bug exists on `master` as well, I'll create an issue for it.

noisysocks commented:
As per the discussion over in https://github.com/WordPress/gutenberg/pull/4500, I've implemented lazy loading in a one-off manner in e51f342. The 1.9 MB of CodeMirror scripts and styles are now not loaded until a HTML block is instantiated.

noisysocks commented:
Rebased and fixed conflicts. How do we feel about getting this in? cc. @westonruter @mtias

westonruter commented on components/code-editor/README.md:
```diff
@@ -0,0 +1,56 @@
+	return (
+		<CodeEditor
+			value={ '<p>This is some <b>HTML</b> code that will have syntax highlighting!</p>' }
+		    onChange={ value => console.log( value ) }
```
Attributes here have inconsistent indentation.

noisysocks replied:
Oops. Fixed in 87af208e500b0dbde3bcb68b0b846a1379692267.

westonruter commented on components/code-editor/README.md:
```diff
@@ -0,0 +1,56 @@
+=======
+
+CodeEditor is a React component that provides the user with a code editor
+that has syntax highliting and linting.
```
`s/highliting/highlighting/`

noisysocks replied:
Oops. Fixed in 87af208e500b0dbde3bcb68b0b846a1379692267.

westonruter commented on components/code-editor/editor.js:
```diff
@@ -0,0 +1,105 @@
+		this.lastCursor = editor.getCursor();
+	}
+
+	onKeyHandled( editor, name, event ) {
```
I love this.

westonruter reviewed (approved):
A couple minor things I commented on.

Also, there is a bit of an unexpected UX behavior whereby if I shift-arrow to select some text, I get the block toolbar appearing. I understand that normally makes sense because toolbar buttons like bold and italic. But since there are no such buttons in the toolbar for Custom HTML (yet, as I suppose there could be shortcuts to add common tags), it is a little distracting to see the HTML/Preview buttons flash whenever I select some text with the keyboard.

westonruter commented:
Per my review comment, here's me with cursor in block:

![image](https://user-images.githubusercontent.com/134745/36567644-5c6b2168-17e4-11e8-8fd7-7ae30844967f.png)

And then as soon as I select some text with keyboard:

![image](https://user-images.githubusercontent.com/134745/36567662-69f0e336-17e4-11e8-8f26-b1827ca02cd8.png)

It doesn't feel like the HTML/Preview buttons show appear.

noisysocks committed:
Fix typos in CodeEditor README

noisysocks commented:
Thanks for reviewing this, @westonruter!

> It doesn't feel like the HTML/Preview buttons show appear.

Good catch. I agree that it feels weird. I think to address this we need to split `<Slot name="Block.Toolbar">` into two slots: one for toolbar buttons that affect the block and one for toolbar buttons that affect the selected text. Only the latter should be shown when text is selected.

Since this would require moving a few things around and it's not a major issue, I'll defer this work for later.
